### ISSUE TITLE
Same prediction values for linear regression while using mxnet in R

### ISSUE THREAD

**Issue by Mrugankakarte** [dmlc/mxnet#5979]
I am new to mxnet and I'm trying to solve a linear regression. My training data contains 1460 observations and 303 variables.

I am using R v3.4.0. Since mxnet is not updated for R v3.4.0 , I have installed mxnet using `install.packages("https://github.com/jeremiedb/mxnet_winbin/raw/master/mxnet.zip", repos = NULL)` as mentioned by jeremiedb in #5948.

I created following frame for the neural network, but the predictions I get have a constant value. I tried different values for batch size, num.rounds but I still get constant values.
When I change the activation type to "Relu" or "softrelu", I get Nan while training. Even using different optimizers is not helping. 
Can anyone please point out where I am going wrong? Thanks in advance!!

`>data <- mx.symbol.Variable("data")`
`>fc1 <- mx.symbol.FullyConnected(data, name="fc1", num_hidden=300)`
`>act1 <- mx.symbol.Activation(fc1, name="sigm1", act_type="sigmoid")`
`>fc2 <- mx.symbol.FullyConnected(act1, name="fc2", num_hidden=300)`
`>act2 <- mx.symbol.Activation(fc2, name="sigm2", act_type="sigmoid")`
`>fc3 <- mx.symbol.FullyConnected(act2, name="fc3", num_hidden=300)`
`>act3 <- mx.symbol.Activation(fc3, name="sigm3", act_type="sigmoid")`
`>fc4 <- mx.symbol.FullyConnected(act3, name="fc4", num_hidden=1)`
`>linear_reg_output <- mx.symbol.LinearRegressionOutput(fc4)`
`>mx.set.seed(0)`
`>model <- mx.model.FeedForward.create(symbol = linear_reg_output,`
                                     `X = train.x, `
                                     `y = train.y, `
                                     `ctx = mx.cpu(),`
                                     `num.round = 200,`
                                     `array.batch.size = 50,`
                                     `learning.rate = 1e-6,`
                                     `eval.data = list(data = valid.x, label = valid.y),`
                                     `#momentum = 0.05,`
                                     `#optimizer = "adam", `
                                     `#optimizer = "rmsprop",`
                                     `eval.metric = mx.metric.rmse)`
`>preds <- predict(model, valid.x)`
`>sqrt(mean((preds-valid.y)^2))`

`> head(preds)`
     `119142.6 119142.6 119142.6 119142.6 119142.6 119142.6 119142.6 119142.6 119142.6 119142.6`

**Comment by Soonhwan-Kwon** [comment]
try mx.random.seed(hash(datetime.now())) on the top of your code. Initialization of parameters are same if there is no random seed.

**Comment by Mrugankakarte** [comment]
Gives error could not find function mx.random.seed

Does this require any library other mxnet? Also, is this not same as using `>mx.set.seed(seed)` ?

**Comment by Soonhwan-Kwon** [comment]
Yes it is . I confused you are using python version. If you put seed any other integer value it may leads you different prediction.

**Comment by Mrugankakarte** [comment]
I'm trying to predict house prices. Same prediction as in I get same predicted values for every predictor.

`head(preds)`
`119142.6 119142.6 119142.6 119142.6 119142.6 119142.6 119142.6 119142.6 119142.6 119142.6`
here preds contains all the predectied values of each test variable.

**Comment by Soonhwan-Kwon** [comment]
maybe after feedforward.create you need model.fit(X=train.y) to train your model.

**Comment by Mrugankakarte** [comment]
`mx.model.feedforward.create()` trains the model.
Issue #4287 has an example comparing nnet, nerualnet and mxnet packages. I have written similar code but for some reason I am not getting the desired output.

**Comment by thirdwing** [comment]
@Mrugankakarte Can you share the data?

**Comment by Mrugankakarte** [comment]
Yeah sure. I'm solving [House prices: Advanced regression techniques](https://www.kaggle.com/c/house-prices-advanced-regression-techniques) problem from kaggle. 
I have transformed the data. All the transformed and original data files with scripts are present [here](https://github.com/Mrugankakarte/Regression/tree/master/Data).

**Comment by Mrugankakarte** [comment]
When I use the following code while creating the network ; it gives the prediction correctly 
`>data <- mx.symbol.Variable("data")`  
`>fc1 <- mx.symbol.FullyConnected(data, name="fc1", num_hidden=1)`  
`>linear_reg_output <- mx.symbol.LinearRegressionOutput(fc1)`

Above code means that it is simple regression where input and output nodes are connected to each other directly, right?
When I use the multi-layer network I get same predictions for each predictor as mentioned earlier

`data <- mx.symbol.Variable("data")`
`fc1 <- mx.symbol.FullyConnected(data, name="fc1", num_hidden=300)`
`act1 <- mx.symbol.Activation(fc1, name="sigm1", act_type="sigmoid")`
`fc2 <- mx.symbol.FullyConnected(act1, name="fc2", num_hidden=300)`
`act2 <- mx.symbol.Activation(fc2, name="sigm2", act_type="sigmoid")`
`fc3 <- mx.symbol.FullyConnected(act2, name="fc3", num_hidden=300)`
`act3 <- mx.symbol.Activation(fc3, name="sigm3", act_type="sigmoid")`
`fc4 <- mx.symbol.FullyConnected(act3, name="fc4", num_hidden=1)`
`linear_reg_output <- mx.symbol.LinearRegressionOutput(fc4)`

**Comment by Liete** [comment]
Hello,

I have the same problem. Each time I use a more sophisticated network, all the predictions (even during training) are the same. The neural net regresses towards the mean of all values and stays there.

This seems similar to issues found by others:
[http://stackoverflow.com/questions/15572091/neuralnet-prediction-returns-the-same-values-for-all-predictions](url)
[https://github.com/dmlc/mxnet/issues/1546](url)
[http://stackoverflow.com/questions/42696418/cant-predict-in-mxnet-0-94-for-r](url)

Here are the things I have tried and/or found out:

- Normalized the labels to be within 0 and 1 (did not solve the issue)
- Added different activation functions before the final fully connected layer with 1 hidden node that feeds into mx.symbol.LinearRegressionOutput() (did not solve the issue, made it worse)
- Simple networks with no activation function work fine, although they also tend not to converge satisfactory during training
- The number of nodes in the layer prior to the final fully connected layer do not matter much (did not solve the issue)
- Different optimizers or different batch sizes used for training do not change this behaviour
- RMSE, MAE or MSE or costum metric functions all result in the same predictions for all samples

Any word on why this occurs or any suggestions how to analyse a continous outcome variable using deep neural networks e.g. using inception or restnet architecture?

**Comment by Liete** [comment]
Alright I figured out why it does not work: I had MXNET_BACKWARD_DO_MIRROR set to 1. I disabled it and now the networks converge on decent values and actually produce different predictions. Hope this was the root of the problem.

**Comment by Mrugankakarte** [comment]
Hi, 
Can you please tell me how exactly you disabled the MXNET_BACKWARD_DO_MIRROR?

Some kind of example with a code will be really helpful.
Thanks

**Comment by Liete** [comment]
Hey,

Just set the environmental variable MXNET_BACKWARD_DO_MIRROR to 0, either in python or outside:
`import os`
`os.environ['MXNET_BACKWARD_DO_MIRROR'] = '1'`

**Comment by Mrugankakarte** [comment]
I tried setting the environment variable MXNET_BACKWARD_DO_MIRROR = 0 but had no luck. I am still getting the same value for predictions. I read the documentation for mxnet; The variable is by default set to 0. [Link](http://mxnet.io/how_to/env_var.html) 

I think there is some problem with the way I am creating the network or maybe with the input data that I am feeding to the network.
Can 'relu' activation function be not used for linear regression? because I get Nan while training when I use it.

**Comment by Liete** [comment]
I had similar experiences with relu and softrelu as well. It did help to not have any activation function between the final layer (with a single node) and the previous fully connected layer. 
 
I was able to train an inception v3 based neural net using a linear regression output and it did quite well for around 100 epochs but then it also started to produce the same prediction for all samples.

**Comment by sagaragrawal13** [comment]
@thirdwing  @Mrugankakarte 

I am also getting same prediction value for binary classification problem when I use RMSPROP optimizer. When I tried using SGD, output is fine. 

In terms of training, running with both optimizers for 50 rounds with learning rate of 0.01. 

Comparison in terms of error metric: 

Optimizer                 Beginning RMSE                  Ending RMSE
rmsprop                     0.4728                                  0.4640
sgd                             0.4542                                  0.4331

Could you please help me out here. Thanks.

**Comment by rcapobianco** [comment]
Similarly, I am getting the same prediction for each input sample. This happens both with MAE regression and softmax in my case, whenever I have multiple layers of ReLUs. The network starts just fine, and optimizer drives it to the setting in which predictions are always the same.

Any ideas?

**Comment by Mrugankakarte** [comment]
Hi,
I tried using `keras` package too but got same results. After changing some parameters, now I am getting different values for different predictor. This was successful in `keras` package, I haven't tried it yet in `mxnet`.

1. Try reducing the batch size and increase epochs so that net can be trained properly.
2. Reduce the size of net. (Change the architecture)
3. Use 'linear' activation function if 'relu' doesn't work.
4. Make sure you have sufficient units for input layer. 
5. Try different learning rates from 0.001 - 0.9.

I will try to do this for my problem (House prices: Advanced regression techniques) mentioned in this issue any post the result images asap.

**Comment by thirdwing** [comment]
@Mrugankakarte 

I just modified your code a little: (1) remove the "id" column; (2) normalization the `train.y`.

Although not a good result, at least not all the same.

```r
library(caret)
library(mxnet)

#Reading the cleaned data
train <- read.csv("transformed train.csv")
test <- read.csv("transformed test.csv")
test1 <- data.matrix(test[, -2])

#Dividing training data into train and validation
ind <- createDataPartition(train$SalePrice, p = 0.6, list = F)
train.x <- data.matrix(train[ind, c(-1,-2)])
train.y <- train[ind, 2]
train.y <- (train.y - 34000) / (760000 - 34000)
valid.x <- data.matrix(train[-ind, c(-1,-2)])
valid.y <- train[-ind, 2]
valid.y <- (valid.y - 34000) / (760000 - 34000)

mx.set.seed(42)
model <- mx.mlp(data = train.x,
                label = train.y,
                hidden_node = 20,
                out_node = 1,
                num.round = 50,
                learning.rate = 1e-6,
                momentum = 0.9,
                eval.metric = mx.metric.rmse,
                out_activation = "rmse")

preds <- predict(model, valid.x)

preds <- preds * (760000 - 34000) + 34000

head(preds[1, ])
## [1] 36875.11 36565.02 35717.43 36714.63 36972.66 38572.08
```

**Comment by antonysamuelb** [comment]
I had this same issue, but for different data.

Initially I tried with a single hidden layer and all outputs gave the mean value.
I scaled the data between 0 and 1 and changed the optimizer from 'sgd' to 'rmsprop'. This improved things a lot.
I could move to two hidden layers now and the results made sense (to an extent).
However when I moved beyond this point I again faced the issue of all outputs having the mean value.
I have now tried initializing different types of weights and that seems to work. specifically the parameter that i used was initializer  = mx.init.normal(0.2) within the mx.model.FeedForward.create() function. 

I am running further tests right now. I will update further on this once I test more number of hidden layers.

**Comment by Liete** [comment]
Hey,

I also had some sucess with a different approach. I fit the model with some very simple data that are similar to the final data I want to use. Next, I loaded the pre-trained model and trained it on the true data and that seemed to work just fine. I was also able to switch between different output metrics without a problem using this approach. So initialization seems to be somewhat crucial.

**Comment by Seager1989** [comment]
I have the same problem on my project as using the symbol method to train an ANN. After normalizing the train.y. The result seems good. This might be caused by the different data scales between the input and output. This might cause the large weights difference and cannot be convergence. You can back to see if the same value is almost equal to the mean value of the train.y.

**Comment by ngupta23** [comment]
I had a similar problem with all outputs predicting the same value. For my case, there were a couple of things that I had to change to fix this. 

Change the architecture of the neural network (number of neurons in the layers). There was no fixed rule that worked for me. In some cases, when I increased the number of neurons, the predictions were same for all observations. In other cases, when I increased the neurons further, the predictions were better.

What also helped was increasing the number of epochs (num.round). I was initially using the default 10, after increasing it to 100 and above, it gave better results. Maybe 10 epochs was not enough to update the weights enough from the random initialization.

Another thing that impacted the results was the learning rate. Decreasing it too  much (1e-5 for my dataset) caused me to get the same predictions for all observations. I had to keep it at around 1e-3 to make it work.

All the above changes were made orthogonally (make change to a single hyperparameter and observe the change in the predictions). It is possible that changing these hyperparameters simultaneously might lead to other conclusions. But the bottom line is that changing the architecture and hyperparaneter values will solve the issue, just that it might take a while to figure out what is the right range for the hyperparameters.
